### PR TITLE
[Merged by Bors] - feat: port Order.RelIso.Basic

### PR DESCRIPTION
**Pull request by ChrisHughes24**
SHA 76171581280d5b5d1e2d1f4f37e5420357bdc636

Finished the port

### CODE REVIEW

File: Mathlib/Order/RelIso/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+  coe_injective' f g h := by
+    cases f
+    cases g
+    congr
+  map_rel := map_rel'
+
+/-- Auxiliary instance if `rel_hom_class.to_fun_like.to_has_coe_to_fun` isn't found -/
+instance : CoeFun (r →r s) fun _ => α → β :=
+  ⟨fun o => o.toFun⟩
+
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:107]: Is this really needed? It seems likely that this is causing the delaboration issue. If we can't remove it, perhaps try adding `instance [coe] RelHom.toFun`.

File: Mathlib/Order/RelIso/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,778 @@
...
+initialize_simps_projections RelHom (toFun → apply)
+
+protected theorem map_rel (f : r →r s) {a b} : r a b → s (f a) (f b) :=
+  f.map_rel'
+#align rel_hom.map_rel RelHom.map_rel
+
+@[simp]
+theorem coe_fn_to_fun (f : r →r s) : (f.toFun : α → β) = f :=
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:115]: I think you should just be writing `(f : \a \to \b)` here, no?

File: Mathlib/Order/RelIso/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+  rfl
+#align rel_hom.coe_fn_to_fun RelHom.coe_fn_to_fun
+
+/-- The map `coe_fn : (r →r s) → (α → β)` is injective. -/
+theorem coe_fn_injective : @Function.Injective (r →r s) (α → β) toFun :=
+  FunLike.coe_injective
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:121]: I wonder if this should/could be `Injective fun (f : r →r s) => (f : α → β)`?

**Comment by semorrison** [comment]
Could you please satisfy the style linter (long lines) and make sure everything is imported in `Mathlib.lean`.

**Comment by Ruben-VandeVelde** [comment]
Can you take a look at the lean lint errors too?
```lean
/- The `docBlame` linter reports:
DEFINITIONS ARE MISSING DOCUMENTATION STRINGS: -/
-- Mathlib.Order.RelIso.Basic
#check RelHom.toFun.{u_1, u_2} /- definition missing documentation string -/
#check RelHom.map_rel'.{u_1, u_2} /- definition missing documentation string -/
#check term_→r_ /- definition missing documentation string -/
#check RelHomClass.map_rel.{u_1, u_2, u_3} /- definition missing documentation string -/
#check RelEmbedding.map_rel_iff'.{u_1, u_2} /- definition missing documentation string -/
#check term_↪r_ /- definition missing documentation string -/
#check RelIso.map_rel_iff'.{u_1, u_2} /- definition missing documentation string -/
#check term_≃r_ /- definition missing documentation string -/

/- The `simpNF` linter reports:
SOME SIMP LEMMAS ARE NOT IN SIMP-NORMAL FORM.
see note [simp-normal form] for tips how to debug this.
https://leanprover-community.github.io/mathlib_docs/notes.html#simp-normal%20form -/
-- Mathlib.Order.RelIso.Basic
#check RelEmbedding.inj.{u_2, u_1} /- simp can prove this:
  by simp only [EmbeddingLike.apply_eq_iff_eq]
One of the lemmas above could be a duplicate.
If that's not the case try reordering lemmas or adding @[priority].
 -/
#check RelIso.eq_iff_eq.{u_2, u_1} /- simp can prove this:
  by simp only [EmbeddingLike.apply_eq_iff_eq]
One of the lemmas above could be a duplicate.
If that's not the case try reordering lemmas or adding @[priority].
 -/

/- The `synTaut` linter reports:
THE FOLLOWING DECLARATIONS ARE SYNTACTIC TAUTOLOGIES. This usually means that they are of the form `∀ a b ... z, e₁ = e₂` where `e₁` and `e₂` are identical expressions. We call declarations of this form syntactic tautologies. Such lemmas are (mostly) useless and sometimes introduced unintentionally when proving basic facts using `rfl`, when elaboration results in a different term than the user intended. You should check that the declaration really says what you think it does. -/
-- Mathlib.Order.RelIso.Basic
#check RelEmbedding.to_rel_hom_eq_coe.{u_2, u_1} /- LHS equals RHS syntactically -/
#check RelEmbedding.coe_fn_to_embedding.{u_2, u_1} /- LHS equals RHS syntactically -/
#check RelIso.to_rel_embedding_eq_coe.{u_2, u_1} /- LHS equals RHS syntactically -/
#check RelIso.coe_coe_fn.{u_2, u_1} /- LHS equals RHS syntactically -/
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+
+end
+
+namespace RelHomClass
+
+variable {F : Type _}
+
+protected theorem is_irrefl [RelHomClass F r s] (f : F) : ∀ [IsIrrefl β s], IsIrrefl α r
+  | ⟨H⟩ => ⟨fun _ h => H _ (map_rel f h)⟩
+#align rel_hom_class.is_irrefl RelHomClass.is_irrefl
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:80]:
```suggestion
protected theorem isIrrefl [RelHomClass F r s] (f : F) : ∀ [IsIrrefl β s], IsIrrefl α r
  | ⟨H⟩ => ⟨fun _ h => H _ (map_rel f h)⟩
#align rel_hom_class.is_irrefl RelHomClass.isIrrefl
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+
+protected theorem is_asymm [RelHomClass F r s] (f : F) : ∀ [IsAsymm β s], IsAsymm α r
+  | ⟨H⟩ => ⟨fun _ _ h₁ h₂ => H _ _ (map_rel f h₁) (map_rel f h₂)⟩
+#align rel_hom_class.is_asymm RelHomClass.is_asymm
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:84]:
```suggestion
protected theorem isAsymm [RelHomClass F r s] (f : F) : ∀ [IsAsymm β s], IsAsymm α r
  | ⟨H⟩ => ⟨fun _ _ h₁ h₂ => H _ _ (map_rel f h₁) (map_rel f h₂)⟩
#align rel_hom_class.is_asymm RelHomClass.isAsymm
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,769 @@
...
+  generalize h : f a = b
+  intro ac
+  induction' ac with _ H IH generalizing a
+  subst h
+  exact ⟨_, fun a' h => IH (f a') (map_rel f h) _ rfl⟩
+#align rel_hom_class.acc RelHomClass.acc
+
+protected theorem well_founded [RelHomClass F r s] (f : F) : ∀ _ : WellFounded s, WellFounded r
+  | ⟨H⟩ => ⟨fun _ => RelHomClass.acc f _ (H _)⟩
+#align rel_hom_class.well_founded RelHomClass.well_founded
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:96]:
```suggestion
protected theorem wellFounded [RelHomClass F r s] (f : F) : ∀ _ : WellFounded s, WellFounded r
  | ⟨H⟩ => ⟨fun _ => RelHomClass.acc f _ (H _)⟩
#align rel_hom_class.well_founded RelHomClass.wellFounded
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,769 @@
...
+    (by
+      refine' RelHomClass.well_founded (RelHom.mk _ _ : s →r r)
+      · exact Classical.choose hf.hasRightInverse
+
+      intro a b h
+      apply o.2
+      convert h
+      iterate 2 apply Classical.choose_spec hf.hasRightInverse)
+    (RelHomClass.well_founded (⟨f, o.1⟩ : r →r s))
+#align surjective.well_founded_iff Surjective.well_founded_iff
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:195]:
```suggestion
theorem Surjective.wellFounded_iff {f : α → β} (hf : Surjective f)
    (o : ∀ {a b}, r a b ↔ s (f a) (f b)) :
    WellFounded r ↔ WellFounded s :=
  Iff.intro
    (by
      refine' RelHomClass.well_founded (RelHom.mk _ _ : s →r r)
      · exact Classical.choose hf.hasRightInverse

      intro a b h
      apply o.2
      convert h
      iterate 2 apply Classical.choose_spec hf.hasRightInverse)
    (RelHomClass.well_founded (⟨f, o.1⟩ : r →r s))
#align surjective.well_founded_iff Surjective.wellFounded_iff
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,769 @@
...
+
+/-- A relation embedding with respect to a given pair of relations `r` and `s`
+is an embedding `f : α ↪ β` such that `r a b ↔ s (f a) (f b)`. -/
+structure RelEmbedding {α β : Type _} (r : α → α → Prop) (s : β → β → Prop) extends α ↪ β where
+  /-- Elements are related iff they are related after apply a `RelEmbedding` -/
+  map_rel_iff' : ∀ {a b}, s (toEmbedding a) (toEmbedding b) ↔ r a b
+#align rel_embedding RelEmbedding
+
+-- mathport name: «expr ↪r »
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:204]:
```suggestion
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 9]
```diff
@@ -0,0 +1,769 @@
...
+#align rel_embedding.preimage RelEmbedding.preimage
+
+theorem eq_preimage (f : r ↪r s) : r = f ⁻¹'o s := by
+  ext (a b)
+  exact f.map_rel_iff.symm
+#align rel_embedding.eq_preimage RelEmbedding.eq_preimage
+
+protected theorem is_irrefl (f : r ↪r s) [IsIrrefl β s] : IsIrrefl α r :=
+  ⟨fun a => mt f.map_rel_iff.2 (irrefl (f a))⟩
+#align rel_embedding.is_irrefl RelEmbedding.is_irrefl
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:317]: Here and below, lots of `isX` instead of `is_X`.

File: Mathlib/Order/RelIso/Basic.lean [hunk 10]
```diff
@@ -0,0 +1,769 @@
...
+  inj' _ _ h := antisymm ((hf _ _).1 (h ▸ refl _)) ((hf _ _).1 (h ▸ refl _))
+  map_rel_iff' := hf _ _
+#align rel_embedding.of_map_rel_iff RelEmbedding.ofMapRelIff
+
+@[simp]
+theorem of_map_rel_iff_coe (f : α → β) [IsAntisymm α r] [IsRefl β s]
+    (hf : ∀ a b, s (f a) (f b) ↔ r a b) :
+    (ofMapRelIff f hf : r ↪r s) = f :=
+  rfl
+#align rel_embedding.of_map_rel_iff_coe RelEmbedding.of_map_rel_iff_coe
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:425]:
```suggestion
theorem ofMapRelIff_coe (f : α → β) [IsAntisymm α r] [IsRefl β s]
    (hf : ∀ a b, s (f a) (f b) ↔ r a b) :
    (ofMapRelIff f hf : r ↪r s) = f :=
  rfl
#align rel_embedding.of_map_rel_iff_coe RelEmbedding.ofMapRelIff_coe
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 11]
```diff
@@ -0,0 +1,769 @@
...
+    · exact asymm (H _ _ h') h
+
+
+#align rel_embedding.of_monotone RelEmbedding.ofMonotone
+
+@[simp]
+theorem of_monotone_coe [IsTrichotomous α r] [IsAsymm β s] (f : α → β) (H) :
+    (@ofMonotone _ _ r s _ _ f H : α → β) = f :=
+  rfl
+#align rel_embedding.of_monotone_coe RelEmbedding.of_monotone_coe
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:449]:
```suggestion
theorem ofMonotone_coe [IsTrichotomous α r] [IsAsymm β s] (f : α → β) (H) :
    (@ofMonotone _ _ r s _ _ f H : α → β) = f :=
  rfl
#align rel_embedding.of_monotone_coe RelEmbedding.ofMonotone_coe
```

File: Mathlib/Order/RelIso/Basic.lean [hunk 12]
```diff
@@ -0,0 +1,769 @@
...
+
+end RelEmbedding
+
+/-- A relation isomorphism is an equivalence that is also a relation embedding. -/
+structure RelIso {α β : Type _} (r : α → α → Prop) (s : β → β → Prop) extends α ≃ β where
+  /-- Elements are related iff they are related after apply a `RelIso` -/
+  map_rel_iff' : ∀ {a b}, s (toEquiv a) (toEquiv b) ↔ r a b
+#align rel_iso RelIso
+
+-- mathport name: «expr ≃r »
```
Comment (semorrison) [inline Mathlib/Order/RelIso/Basic.lean:536]:
```suggestion
```

**Comment by semorrison** [comment]
bors d+

**Review by Ruben-VandeVelde — APPROVED** [review]
This is good to go, IMO

**Comment by semorrison** [comment]
bors merge

**Comment by alexjbest** [comment]
Sorry but what is SHA `19091d8f2e537e76a81d1608e011dd4d03f94642` @ChrisHughes24 it makes my `./scripts/port_status.py` complain that `19091d8f2e537e76a81d1608e011dd4d03f94642..HEAD` is not a valid range

**Comment by ChrisHughes24** [comment]
Sorry I made a mistake and used the commit hash for the mathlib3port repo. The SHA should be 76171581280d5b5d1e2d1f4f37e5420357bdc636 There was one other PR where I made this mistake. I'll check it

**Pull request closed without merging**